Engine/src/ECS/Systems/BaseSystem.bf

```diff
@@ -0,0 +1,318 @@
+				{
+					delete entityComponents;
+				}
+				return false;
+			}
+
+			if (!component.IsInitialized)
+			{
+				_uninitializedComponents.Enqueue(component);
+			}
+			if (isNewEntity)
+			{
+				EntityToComponents[parent.Id] = entityComponents;
+			}
+			entityComponents.Add(component);
+			component.[Friend]IsQueuedForDeletion = false;
+			_entityRegistrationChecks.Add(parent.Id);
+			return true;
+		}
+
+		protected bool CanBeRegistered(BaseComponent component)
+		{
+			for (let type in _registeredTypes)
+			{
+				if (CanBeRegisteredAsType(component, type))
+				{
+					return true;
+				}
+			}
+			return false;
+		}
+
+		protected bool CanBeRegisteredAsType(BaseComponent component, Type type)
+		{
+			let componentType = component.GetType();
+			if (componentType.TypeId == type.TypeId || componentType.IsSubtypeOf(type))
+			{
+				return true;
+			}
+			return false;
+		}
+
+		protected void ClearEmptyEntities()
+		{
+			let entitiesToRemove = new List<EntityId>();
+			defer delete entitiesToRemove;
+
+			for (let item in EntityToComponents)
+			{
+				if (item.value == null || item.value.Count == 0)
+				{
+					entitiesToRemove.Add(item.key);
+				}
+			}
+			for (let entity in entitiesToRemove)
+			{
+				EntityToComponents.Remove(entity);
+			}
+		}
+
+		protected virtual void Draw()
+		{
+			if (!IsEnabled)
+			{
+				return;
+			}
+			InitializeComponents();
+
+			for (let item in EntityToComponents)
+			{
+				Draw(item.key, item.value);
+			}
+		}
+
+		protected virtual void Draw(EntityId entityId, List<BaseComponent> components)
+		{
+
+		}
+
+		protected List<BaseComponent> GetComponentsOfEntity(Entity entity)
+		{
+			List<BaseComponent> entityComponents = ?;
+			EntityToComponents.TryGetValue(entity.Id, out entityComponents);
+			return entityComponents;
+		}
+
+		protected virtual Result<void, InitializationError> Initialize()
+		{
+			if (!IsEnabled || IsInitialized)
+			{
+				return .Err(.AlreadyInitialized);
+			}
+			if (InitializeComponents() == .Err)
+			{
+				return .Err(.Unknown);
+			}
+			IsInitialized = true;
+			return .Ok;
+		}
+
+		protected virtual Result<void> Initialize(BaseComponent component)
+		{
+			component.[Friend]IsInitialized = true;
+			return .Ok;
+		}
+
+		protected Result<void> InitializeComponents()
+		{
+			while (_uninitializedComponents.Count != 0)
+			{
+				let component = _uninitializedComponents.Dequeue();
+				if (!component.IsInitialized)
+				{
+					if (Initialize(component) == .Err)
+					{
+						return .Err;
+					}
+				}
+			}
+			return .Ok;
+		}
+
+		protected virtual void PostUpdate()
+		{
+			ClearEmptyEntities();
+		}
+
+		protected virtual void PreUpdate()
+		{
+			ClearEmptyEntities();
+		}
+
+		protected bool RefreshEntityRegistration(EntityId entityId)
+		{
+			List<BaseComponent> components = ?;
+			if (!EntityToComponents.TryGetValue(entityId, out components))
+			{
+				return false;
+			}
+
+			let componentsToRemove = new List<BaseComponent>();

```

Comment:
`componentsToRemove` is never deleted, i suggest using `new:ScopedAlloc!` which will delete allocated object at the end of the scope.

Reply:
When I used `new:ScopedAlloc!`, I ran into more memory leaks? I think I was likely messing it up or something. I went ahead and added a `defer delete` statement instead.